### PR TITLE
Get rid of pickle for fluid inference

### LINKED ISSUE #7221: Get rid of pickle for Storing format of inference
Currently Pickle is used in fluid.io.save_inference_model to save inference protobuf string and feed and fetch var names.

However it is not well supported for the c++ side to parse the serializing format provided by pickle. 

To make the inference API concise, we need to get rid of pickle.

### PR DESCRIPTION
**Pull request by kexinzhao**
fixes #7221 

Put fetch_var_names and feed_var_names as repeated string fields in the ProgramDesc

To test this PR：

1. build paddle based on this pr
2. Use pip install -U /home/liuyiqun01/PaddlePaddle/Paddle/build_paddle/dist/opt/paddle/share/wheels/paddlepaddle_gpu-0.11.0-cp27-cp27mu-linux_x86_64.whl to install the latest Paddle and run python test_recognize_digits_mlp.py to train the model, then you will get the inference model in sub-directory recognize_digits_mlp.inference.model

After these steps, you can run the example using:
```
$ GLOG_v=3 ./paddle/inference/example --dirname=/home/liuyiqun01/PaddlePaddle/Paddle/python/paddle/v2/fluid/tests/book/recognize_digits_mlp.inference.model 
```

The output is similar to #7097

### CODE REVIEW

File: paddle/framework/framework.proto [hunk 1]
```diff
@@ -143,4 +143,8 @@ message BlockDesc {
 // Please refer to
 // https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/program.md
 // for more details.
-message ProgramDesc { repeated BlockDesc blocks = 1; }
+message ProgramDesc {
+  repeated BlockDesc blocks = 1;
+  repeated string feed_var_names = 2;
```
Comment (wangkuiyi) [inline paddle/framework/framework.proto]: Why do we need to add feed_var_names and fetch_var_names here?  The current ProgramDesc, which supports training, should be able to read training data — maybe from some data feeding operators I vaguely remember?  If so, it seems that we can reuse these feeding operators, or invent some new operators for data feeding specifically at inference time?

Comment (wangkuiyi) [reply paddle/framework/framework.proto]: BTW, given the fact that we are moving to use a WhileOp to represent the iteration loop at training time, I don’t expect a ProgramDesc that describes a training program can be shared and reused without any modification at inference time.  As we’d need a new ProgramDesc specifically for inference, we can have inference data feeding operators in this new ProgramDesc.

Comment (kavyasrinet) [reply paddle/framework/framework.proto]: @wangkuiyi This is the approach we have been discussing too here: https://github.com/PaddlePaddle/Paddle/issues/7328 . Defining a ProgramDesc for Inference separately that has all Inference specific requirements addressed separately, and that can reuse our already existing ProgramDesc.

Comment (kexinzhao) [reply paddle/framework/framework.proto]: Thanks @wangkuiyi for the clarification on ProgramDesc. We are exploring different paths of obtaining a Desc for inference and discussing which one to use. Based on your comment, I think we will follow things the proposal in https://github.com/PaddlePaddle/Paddle/issues/7328. @abhinavarora is working on a PR based on that.

Comment (Xreki) [reply paddle/framework/framework.proto]: There is no `feed_op` or `fetch_op` in current train and inference `ProgramDesc`. Each time we run a train `ProgramDesc`, the `Executor` will clone a copy of the train `ProgramDesc` and insert `feed_op`s and `fetch_op`s to the copy of the train `ProgramDesc`.
@wangkuiyi do you mean to insert `feed_op` and `fetch_op` to inference `ProgramDesc` when saving? It is a way to solve this problem.
There is a proposal of mine, with no need to modify the current `ProgramDesc` or define another `ProgramDesc`, but I am not sure if it works. @kexinzhao @abhinavarora @kavyasrinet 
- We can traverse all the non-persistable variables in the inference `ProgramDesc`. If the variable is not an output of any operators, we think it is a feed variable. If the variable is not an input of any operators, we think it is a fetch variable.

**Pull request closed without merging**